{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "num of mp   9614\n",
      "num of go   15803\n",
      "num of uberon    16242\n",
      "num of intersection     8751\n",
      "num of union      17417\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\ngenerate the gene and disease associations based on mouse gene and mouse diseases\\nassociate the features with gene and diseases\\nwrite the gene and disease out according to the specified data format\\n'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#  associate the human genes\n",
    "\n",
    "# associate the mp\n",
    "# associate the go\n",
    "# associate the uberon\n",
    "\n",
    "\n",
    "# define a general method that could generate each type of association\n",
    "\n",
    "\n",
    "# associate the human disease\n",
    "\n",
    "# select the gene and disease combinations\n",
    "\n",
    "\n",
    "# generate the features for the gene and disease\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pickle as pkl\n",
    "\n",
    "\n",
    "options=[\"intersection\",\"go\",\"uberon\",\"mgi\",\"union\"]\n",
    "\n",
    "# generate the human disease and hp association\n",
    "dis_phe=dict()\n",
    "with open(\"../phenotype_annotation.tab\",\"r\") as f:\n",
    "    for line in f.readlines():\n",
    "        data=line.split(\"\\t\")\n",
    "        if (data[0]==\"OMIM\")&(data[5][:4]==\"OMIM\"):\n",
    "            try:\n",
    "                dis_phe[data[5].strip()].append(data[4].strip())\n",
    "            except:\n",
    "                dis_phe[data[5].strip()]=[data[4].strip()]\n",
    "\n",
    "\n",
    "# obtain the mouse gene and mouse disease association\n",
    "# obtain the human gene and human disease association\n",
    "\n",
    "\n",
    "mgi_do=pd.read_table(\"../MGI_DO.rpt.txt\")\n",
    "human_gene_disease=[]\n",
    "\n",
    "for index in mgi_do.index:\n",
    "    if (mgi_do.loc[index,\"Common Organism Name\"]==\"human\"):\n",
    "        disease=mgi_do.loc[index,\"OMIM IDs\"]\n",
    "        sub_dis=disease.split(\"|\")\n",
    "        human_gene=mgi_do.loc[index,\"EntrezGene ID\"]\n",
    "        if disease:\n",
    "            if human_gene:\n",
    "                for dis in sub_dis:\n",
    "                    human_gene_disease.append([int(human_gene),dis.strip()])\n",
    "\n",
    "mouse_gene_disease=dict()\n",
    "for index in mgi_do.index:\n",
    "    mouse_gene=mgi_do.loc[index,\"Mouse MGI ID\"]\n",
    "    if str(mouse_gene)!=\"nan\":\n",
    "        disease=str(mgi_do.loc[index,\"OMIM IDs\"]).strip()\n",
    "        sub_dis=disease.split(\"|\")\n",
    "        for dis in sub_dis:\n",
    "            if dis !=\"nan\":\n",
    "                try:\n",
    "                    mouse_gene_disease[mouse_gene].append(dis.strip())\n",
    "                except:\n",
    "                    mouse_gene_disease[mouse_gene]=[dis.strip()]\n",
    "\n",
    "## generate the dictionary from human to mouse and from mouse to human\n",
    "human_to_mouse=dict()\n",
    "mouse_to_human=dict()\n",
    "geneName_to_id=dict()\n",
    "with open(\"../HMD_HumanPhenotype.rpt.txt\",'r') as f:\n",
    "    for line in f.readlines():\n",
    "        data=line.split(\"\\t\")\n",
    "        gene_name=data[0].strip()\n",
    "        human_id=data[1].strip()\n",
    "        mouse_id=data[5].strip()\n",
    "        human_to_mouse[human_id]=mouse_id\n",
    "        mouse_to_human[mouse_id]=human_id\n",
    "        geneName_to_id[gene_name]=human_id\n",
    "\n",
    "\n",
    "# obtain the human gene and gene function association\n",
    "\n",
    "\n",
    "gene_go_feature=dict()\n",
    "gene_expression_name=set()\n",
    "with open(\"../goa_human.gaf\",\"r\") as f:\n",
    "    for line in f.readlines():\n",
    "        data=line.split(\"\\t\")\n",
    "        gene_name=data[2].strip()\n",
    "        evidence_score=data[6].strip()\n",
    "        go_id=data[4].split()\n",
    "        if gene_name in geneName_to_id.keys():\n",
    "            if not ((evidence_score==\"IEA\") or (evidence_score==\"ND\")):\n",
    "                human_gene=geneName_to_id[gene_name]\n",
    "                if (human_gene in mouse_to_human.values()):\n",
    "\n",
    "                    try:\n",
    "                        gene_go_feature[human_gene].append(go_id)\n",
    "                    except:\n",
    "                        gene_go_feature[human_gene]=[go_id]\n",
    "\n",
    "\n",
    "# obtain the mammalian phenotypes and gene association\n",
    "\n",
    "mp_pheno=pd.read_table(\"../MGI_GenePheno.rpt.txt\",names=[\"Allelic_Composition\",\"Allele_Symbol\",\"Allele_ID\",\"Genetic_Background\",\"Mammalian_Phenotype_ID\",\"PubMed_ID\",\"MGI_Marker\",\"MGI_Genotype_Accession\"])\n",
    "gene_mp_feature=dict()\n",
    "for index in mp_pheno.index:\n",
    "    mouse_gene=mp_pheno.loc[index,\"MGI_Marker\"].strip()\n",
    "    mouse_pheno=mp_pheno.loc[index,\"Mammalian_Phenotype_ID\"].strip()\n",
    "    if (mouse_gene in mouse_to_human.keys()):\n",
    "        human_gene=str(int(mouse_to_human[mouse_gene]))\n",
    "        try:\n",
    "            gene_mp_feature[human_gene].append(mouse_pheno)\n",
    "        except:\n",
    "            gene_mp_feature[human_gene]=[mouse_pheno]\n",
    "\n",
    "# obtain the gene and uberon associations\n",
    "gene_expression=pd.read_table(\"../E-MTAB-5214-query-results.tpms-2.tsv\")\n",
    "gene_expression=gene_expression.drop(columns=\"EBV-transformed lymphocyte\")\n",
    "gene_expression=gene_expression.drop(columns=\"transformed skin fibroblast\")\n",
    "\n",
    "anatomy_uberon={'Brodmann (1909) area 24': 'UBERON_0006101',\n",
    " 'Brodmann (1909) area 9': 'UBERON_0013540',\n",
    " 'C1 segment of cervical spinal cord': 'UBERON_0006469',\n",
    " 'adrenal gland': 'UBERON_0002369',\n",
    " 'amygdala': 'UBERON_0001876',\n",
    " 'aorta': 'UBERON_0000947',\n",
    " 'atrium auricular region': 'UBERON_0006618',\n",
    " 'breast': 'UBERON_0000310',\n",
    " 'caudate nucleus': 'UBERON_0001873',\n",
    " 'cerebellar hemisphere': 'UBERON_0002245',\n",
    " 'cerebral cortex': 'UBERON_0000956',\n",
    " 'coronary artery': 'UBERON_0001621',\n",
    " 'cortex of kidney': 'UBERON_0001225',\n",
    " 'ectocervix': 'UBERON_0012249',\n",
    " 'endocervix': 'UBERON_0000458',\n",
    " 'esophagogastric junction': 'UBERON_0007650',\n",
    " 'esophagus mucosa': 'UBERON_0002469',\n",
    " 'esophagus muscularis mucosa': 'UBERON_0004648',\n",
    " 'fallopian tube': 'UBERON_0003889',\n",
    " 'greater omentum': 'UBERON_0005448',\n",
    " 'heart left ventricle': 'UBERON_0002084',\n",
    " 'hypothalamus': 'UBERON_0001898',\n",
    " 'lower leg skin': 'UBERON_0004264',\n",
    " 'minor salivary gland': 'UBERON_0001830',\n",
    " 'nucleus accumbens': 'UBERON_0001882',\n",
    " 'ovary': 'UBERON_0000992',\n",
    " 'pancreas': 'UBERON_0001264',\n",
    " 'pituitary gland': 'UBERON_0000007',\n",
    " 'prostate gland': 'UBERON_0002367',\n",
    " 'putamen': 'UBERON_0001874',\n",
    " 'sigmoid colon': 'UBERON_0001159',\n",
    " \"small intestine Peyer's patch\": 'UBERON_0003454',\n",
    " 'stomach': 'UBERON_0000945',\n",
    " 'subcutaneous adipose tissue': 'UBERON_0002190',\n",
    " 'substantia nigra': 'UBERON_0002038',\n",
    " 'suprapubic skin': 'UBERON_0036149',\n",
    " 'testis': 'UBERON_0000473',\n",
    " 'thyroid gland': 'UBERON_0002046',\n",
    " 'tibial artery': 'UBERON_0007610',\n",
    " 'tibial nerve': 'UBERON_0001323',\n",
    " 'transverse colon': 'UBERON_0001157',\n",
    " 'urinary bladder': 'UBERON_0001255',\n",
    " 'uterus': 'UBERON_0000995',\n",
    " 'vagina': 'UBERON_0000996',\n",
    "  'blood':\"UBERON_0000178\",\n",
    "  'liver':\"UBERON_0002107\",\n",
    "  'lung':\"UBERON_0002048\",\n",
    "  \"spleen\":\"UBERON_0002106\",\n",
    "  \"cerebellum\":\"UBERON_0002037\",\n",
    "  \"skeletal muscle tissue\":\"UBERON_0001134\",\n",
    "  \"hippocampus proper\":\"UBERON_0002305\"}\n",
    "\n",
    "\n",
    "# convert the nan value into 0 value\n",
    "for index in gene_expression.index:\n",
    "    for name in gene_expression.columns[2:]:\n",
    "        if (str(gene_expression.loc[index,name])==\"nan\"):\n",
    "            gene_expression.loc[index,name]=0\n",
    "\n",
    "\n",
    "# now filter the uberon that has expression value more than 4.0\n",
    "\n",
    "gene_uberon_feature=dict()\n",
    "threshold=4.0\n",
    "for index in gene_expression.index:\n",
    "    name=gene_expression.loc[index,\"Gene Name\"]\n",
    "    if name in geneName_to_id.keys():\n",
    "        name=geneName_to_id[name]\n",
    "        if name in human_to_mouse.keys():\n",
    "            for column in gene_expression.columns[2:]:\n",
    "                if gene_expression.loc[index,column]>=threshold:\n",
    "                    try:\n",
    "                        gene_uberon_feature[name].add(anatomy_uberon[column])\n",
    "                    except:\n",
    "                        temp_set=set()\n",
    "                        temp_set.add(anatomy_uberon[column])\n",
    "                        gene_uberon_feature[name]=temp_set\n",
    "\n",
    "\"\"\"\n",
    "now we have the data set of uberon, mammalian phenotypes and gene function\n",
    "gene_go_feature\n",
    "gene_mp_feature\n",
    "gene_uberon_feature\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# generate the intersection data\n",
    "gene_intersection_feature=dict()\n",
    "for data in gene_go_feature.keys():\n",
    "    if data in gene_mp_feature.keys():\n",
    "        if data in gene_uberon_feature.keys():\n",
    "            features=set()\n",
    "            # print(gene_go_feature[data])\n",
    "\n",
    "            for value in gene_go_feature[data]:\n",
    "                features.add(value[0])\n",
    "            for value in gene_mp_feature[data]:\n",
    "                features.add(value[0])\n",
    "            for value in gene_uberon_feature[data]:\n",
    "                features.add(value[0])\n",
    "            gene_intersection_feature[data]=features\n",
    "\n",
    "# generate the union data\n",
    "gene_total=set()\n",
    "for data in gene_go_feature.keys():\n",
    "    gene_total.add(data)\n",
    "for data in gene_uberon_feature.keys():\n",
    "    gene_total.add(data)\n",
    "for data in gene_mp_feature.keys():\n",
    "    gene_total.add(data)\n",
    "\n",
    "gene_union_feature=dict()\n",
    "for data in gene_total:\n",
    "    features = set()\n",
    "    if data in gene_go_feature.keys():\n",
    "        for value in gene_go_feature[data]:\n",
    "            features.add(value[0])\n",
    "    if data in gene_mp_feature.keys():\n",
    "        for value in gene_mp_feature[data]:\n",
    "            features.add(value[0])\n",
    "    if data in gene_uberon_feature.keys():\n",
    "        for value in gene_uberon_feature[data]:\n",
    "            features.add(value[0])\n",
    "    gene_union_feature[data] = features\n",
    "\n",
    "print(\"num of mp  \",str(len(gene_mp_feature)))\n",
    "print(\"num of go  \",str(len(gene_go_feature)))\n",
    "print(\"num of uberon   \",str(len(gene_uberon_feature)))\n",
    "print(\"num of intersection    \",str(len(gene_intersection_feature)))\n",
    "print(\"num of union     \",str(len(gene_union_feature)))\n",
    "\n",
    "\n",
    "'''\n",
    "generate the gene and disease associations based on mouse gene and mouse diseases\n",
    "associate the features with gene and diseases\n",
    "write the gene and disease out according to the specified data format\n",
    "'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate the intersection data\n",
    "gene_intersection_feature=dict()\n",
    "for data in gene_go_feature.keys():\n",
    "    if data in gene_mp_feature.keys():\n",
    "        if data in gene_uberon_feature.keys():\n",
    "            features=set()\n",
    "            # print(gene_go_feature[data])\n",
    "\n",
    "            for value in gene_go_feature[data]:\n",
    "                features.add(value[0])\n",
    "            for value in gene_mp_feature[data]:\n",
    "                features.add(value)\n",
    "            for value in gene_uberon_feature[data]:\n",
    "                features.add(value)\n",
    "            gene_intersection_feature[data]=features\n",
    "\n",
    "# generate the union data\n",
    "gene_total=set()\n",
    "for data in gene_go_feature.keys():\n",
    "    gene_total.add(data)\n",
    "for data in gene_uberon_feature.keys():\n",
    "    gene_total.add(data)\n",
    "for data in gene_mp_feature.keys():\n",
    "    gene_total.add(data)\n",
    "\n",
    "gene_union_feature=dict()\n",
    "for data in gene_total:\n",
    "    features = set()\n",
    "    if data in gene_go_feature.keys():\n",
    "        for value in gene_go_feature[data]:\n",
    "            features.add(value[0])\n",
    "    if data in gene_mp_feature.keys():\n",
    "        for value in gene_mp_feature[data]:\n",
    "            features.add(value)\n",
    "    if data in gene_uberon_feature.keys():\n",
    "        for value in gene_uberon_feature[data]:\n",
    "            features.add(value)\n",
    "    gene_union_feature[data] = features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_features(mouse_gene_disease_association,gene_feature,disease_feature,type):\n",
    "    file=open(\"data/\"+type+\"_train.txt\",\"w\")\n",
    "    disease_gene=dict()\n",
    "    total_gene=set()\n",
    "    count=0\n",
    "    for key in mouse_gene_disease_association.keys():\n",
    "        try:\n",
    "            human_gene=mouse_to_human[key]\n",
    "            g_features=gene_feature[human_gene]\n",
    "            diseases=mouse_gene_disease_association[key]\n",
    "            for dis in diseases:\n",
    "                count+=1\n",
    "                d_features=disease_feature[dis]\n",
    "                for data in d_features:\n",
    "                    file.write(disease+' '+\"<http://purl.obolibrary.org/obo/\"+str(data)+\">\"+\"\\n\")\n",
    "                for data in g_features:\n",
    "                    file.write(human_gene+\" \"+'<http://purl.obolibrary.org/obo/'+str(data)+\">\"+\"\\n\")\n",
    "\n",
    "                try:\n",
    "                    disease_gene[dis].append(human_gene)\n",
    "                except:\n",
    "                    disease_gene[dis]=[human_gene]\n",
    "            total_gene.add(human_gene)\n",
    "\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "    with open(\"data/\"+type+\"_disease_gene.pkl\",\"wb\") as f:\n",
    "        pkl.dump(disease_gene,f)\n",
    "    with open(\"data/\"+type+\"_gene_set.pkl\",\"wb\") as f:\n",
    "        pkl.dump(total_gene,f)\n",
    "    print(count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3414\n",
      "3409\n",
      "3253\n",
      "3215\n",
      "3433\n"
     ]
    }
   ],
   "source": [
    "generate_features(mouse_gene_disease,gene_mp_feature,dis_phe,\"mp\")\n",
    "generate_features(mouse_gene_disease,gene_go_feature,dis_phe,\"go\")\n",
    "generate_features(mouse_gene_disease,gene_uberon_feature,dis_phe,\"uberon\")\n",
    "generate_features(mouse_gene_disease,gene_intersection_feature,dis_phe,\"intersection\")\n",
    "generate_features(mouse_gene_disease,gene_union_feature,dis_phe,\"union\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
